stream-chat-android-client/src/main/java/io/getstream/chat/android/client/parser2/CustomObjectDtoAdapter.kt

```diff
@@ -0,0 +1,90 @@
+package io.getstream.chat.android.client.parser2
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import kotlin.reflect.KCallable
+import kotlin.reflect.KClass
+import kotlin.reflect.full.declaredMemberProperties
+
+/**
+ * Base class for implementing Moshi adapters that support our API's dynamic
+ * JSON models.
+ */
+internal abstract class CustomObjectDtoAdapter<Value : Any>(private val kClass: KClass<Value>) {
+
+    private companion object {
+        private const val EXTRA_DATA = "extraData"
+    }
+
+    /**
+     * Names of the declared properties that are inside the [Value] type being
+     * handled. These will not be copied into extraData when parsing with
+     * [parseWithExtraData].
+     */
+    private val memberNames: List<String> by lazy {
+        kClass.declaredMemberProperties.map(KCallable<*>::name).minus(EXTRA_DATA)
+    }
+
+    /**
+     * Moves all values in the input JSON that are not declared properties of
+     * [Value] into an extraValue field, and then parses a [Value] instance
+     * from this transformed data.
+     */
+    protected fun parseWithExtraData(
+        jsonReader: JsonReader,
+        mapAdapter: JsonAdapter<MutableMap<String, Any>>,
+        valueAdapter: JsonAdapter<Value>,
+    ): Value {
+        // Parse full JSON content as a MutableMap
+        val map = mapAdapter.fromJson(jsonReader)!!
+
+        val extraData = mutableMapOf<String, Any>()
+
+        // Save the value of the literal "extraData" field at the root of the object, if present
+        map[EXTRA_DATA]?.let { explicitExtraData ->
+            extraData[EXTRA_DATA] = explicitExtraData
+        }
+
+        // Save the values of non-member fields as extra data
+        map.forEach { entry ->
```

Comment:
`map.filterKeys { key -> key !in memberNames }.let(extraData::putAll)`

Can be an interesting option if you want to avoid `forEach`

Reply:
`forEach` in this case avoids creating an intermediate collection. This serialization code will run often, we're already double-parsing everything due to the transformations required, so I want to keep performance as tight as possible otherwise.